ignite/engine/engine.py

```diff
@@ -123,28 +123,35 @@ def compute_mean_std(engine, batch):
     _state_dict_one_of_opt_keys = ("iteration", "epoch")
 
     def __init__(self, process_function: Callable):
-        self._event_handlers = defaultdict(list)  # type: Dict[Any, List]
+        super(Engine, self).__init__()
         self.logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
         self._process_function = process_function
-        self.last_event_name = None  # type: Optional[Events]
         self.should_terminate = False
         self.should_terminate_single_epoch = False
-        self.state = State()
+        self._state = State(engine=self)
         self._state_dict_user_keys = []  # type: List[str]
-        self._allowed_events = []  # type: List[EventEnum]
 
         self._dataloader_iter = None  # type: Optional[Iterator[Any]]
         self._init_iter = []  # type: List[int]
 
-        self.register_events(*Events)
+        self.register_events(*Events, attr_to_events=State.attr_to_events)
 
         if self._process_function is None:
             raise ValueError("Engine must be given a processing function in order to run.")
 
         _check_signature(process_function, "process_function", self, None)
 
+    @property
+    def state(self) -> State:
+        return self._state
+
+    @state.setter
+    def state(self, new_state: State) -> None:
+        self._state = new_state
+        self._state.engine = self
+
     def register_events(
-        self, *event_names: Union[List[str], List[EventEnum]], event_to_attr: Optional[dict] = None
+        self, *event_names: Union[List[str], List[EventEnum]], attr_to_events: Optional[dict] = None
```

Comment:
You can not remove an arg like that without deprecating it

Reply:
I'm thinking about leaving both for now `event_to_attr` and `attr_to_events`, and do something like this inside the method
```py
if event_to_attr is not None:
        # raise warning about deprecation
        # create attr_to_events from event_to_attr

# then continue the same as now
```